README.md
[![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-2.1-4baaaa.svg)](CODE_OF_CONDUCT.md) 
# radiobrowser-api-rust

## What is radiobrowser-api-rust?
In short it is an API for an index of web streams (audio and video). Streams can be added and searched by any user of the API.

There is an official deployment of this software that is also freely usable at https://api.radio-browser.info

## Features
* Open source
* Freely licensed
* Well documented API
* Automatic regular online checking of streams
* Highliy configurable
* Easy setup for multiple configurations (native, deb-packages, docker, ansible)
* Implemented in Rust-lang
* Multiple request types: query, json, x-www-form-urlencoded, form-data
* Multiple output types: xml, json, m3u, pls, xspf, ttl, csv
* Optional: multi-server setup with automatic mirroring
* Optional: response caching in internal or external cache (redis, memcached)

## Setup

You can do a native setup or a docker setup

### easy all-in-one docker setup with automatic TLS from Let's Encrypt

#### install

* This has been tested on Ubuntu 20.04 LTS
* Automatic redirect from HTTP to HTTPS
* Automatic generation and update of Let's Encrypt certificates
* Automatic start on reboot
* Automatic fetch of station changes and check information from main server
* A+ rating on https://www.ssllabs.com/ssltest/

```bash
# create radiobrowser directory
mkdir -p /srv/radiobrowser
cd /srv/radiobrowser
# download docker-compose file
wget https://raw.githubusercontent.com/segler-alex/radiobrowser-api-rust/stable/docker-compose-traefik.yml -O docker-compose-traefik.yml
wget https://raw.githubusercontent.com/segler-alex/radiobrowser-api-rust/stable/traefik-dyn-config.toml -O traefik-dyn-config.toml
# create database save directory
mkdir -p dbdata
# create ssl certificate cache file
touch acme.json
chmod 0600 acme.json
# install docker (ubuntu)
apt install -qy docker.io
docker swarm init
# set email and domain, they are needed for automatic certificate generation and for the reverse proxy that is included in the package
export SOURCE="my.domain.org"
export EMAIL="mymail@mail.com"
# OPTIONAL: enable checking of stations, this does check all stations once every 24 hours
export ENABLE_CHECK="true"
# deploy app stack
docker stack deploy -c docker-compose-traefik.yml rb

# For security reasons you should also install fail2ban to secure SSH and enable a firewall
apt-get install fail2ban -y
ufw allow ssh
ufw allow http
ufw allow https
ufw enable
```

#### upgrade

```bash
# update compose file
cd /srv/radiobrowser
wget https://raw.githubusercontent.com/segler-alex/radiobrowser-api-rust/stable/docker-compose-traefik.yml -O docker-compose-traefik.yml
wget https://raw.githubusercontent.com/segler-alex/radiobrowser-api-rust/stable/traefik-dyn-config.toml -O traefik-dyn-config.toml
# set email and domain, they are needed for automatic certificate generation and for the reverse proxy that is included in the package
export SOURCE="my.domain.org"
export EMAIL="mymail@mail.com"
# OPTIONAL: enable checking of stations, this does check all stations once every 24 hours
export ENABLE_CHECK="true"
# deploy app stack, old versions will automatically be upgraded
docker stack deploy -c docker-compose-traefik.yml rb
```

### install from distribution package

* download latest tar.gz package (<https://github.com/segler-alex/radiobrowser-api-rust/releases>)
* untar
* configure mysql/mariadb on your machine
* create database and database user
* call install script
* change /etc/radiobrowser/config.toml if needed
* start systemd service

```bash
# download distribution
mkdir -p radiobrowser
cd radiobrowser
wget https://github.com/segler-alex/radiobrowser-api-rust/releases/download/0.7.24/radiobrowser-dist.tar.gz
tar -zxf radiobrowser-dist.tar.gz

# config database
sudo apt install default-mysql-server
cat init.sql | mysql

# install
./install.sh
sudo systemctl enable radiobrowser
sudo systemctl start radiobrowser
```

### debian/ubuntu package

* download latest deb package (<https://github.com/segler-alex/radiobrowser-api-rust/releases>)
* install it
* configure mysql/mariadb
* create database and database user

```bash
wget https://github.com/segler-alex/radiobrowser-api-rust/releases/download/0.7.24/radiobrowser-api-rust_0.7.24_amd64.deb
sudo apt install default-mysql-server
sudo dpkg -i radiobrowser-api-rust_0.7.24_amd64.deb
cat /usr/share/radiobrowser/init.sql | mysql
```

### native setup from source

Requirements:

* rust, cargo (<http://www.rustup.sh)>
* mariadb or mysql

```bash
# install packages (ubuntu 18.04)
curl https://sh.rustup.rs -sSf | sh
sudo apt install libssl-dev pkg-config gcc
sudo apt install default-mysql-server
```

```bash
# clone repository
git clone https://github.com/segler-alex/radiobrowser-api-rust ~/radio

# setup database, compile, install
cd ~/radio
cat init.sql | mysql
bash install_from_source.sh

# test it
xdg-open http://localhost/webservice/xml/countries
# or just open the link with your favourite browser
```

### docker setup

```bash
# start db and api server
docker-compose up --abort-on-container-exit
```

### docker from registry
The repository is at https://hub.docker.com/repository/docker/segleralex/radiobrowser-api-rust
```bash
# !!! DO NOT USE THE FOLLOWING FOR PRODUCTION !!!
# It is just for a quickstart and is a minimal setup.

# create virtual network for communication between database and backend
docker network create rbnet
# start database container
docker run \
    --name dbserver \
    --detach \
    --network rbnet \
    --rm \
    -e MYSQL_DATABASE=radio \
    -e MYSQL_USER=radiouser \
    -e MYSQL_PASSWORD=password \
    -e MYSQL_RANDOM_ROOT_PASSWORD=true \
    -p 3306:3306 \
    mariadb --character-set-server=utf8mb4 --collation-server=utf8mb4_unicode_ci
# start radiobrowser container
docker pull segleralex/radiobrowser-api-rust:0.7.24
docker run \
    --name radiobrowserapi \
    --detach \
    --network rbnet \
    --rm \
    -e DATABASE_URL=mysql://radiouser:password@dbserver/radio \
    -e HOST=0.0.0.0 \
    -p 8080:8080 \
    segleralex/radiobrowser-api-rust:0.7.24 radiobrowser-api-rust -vvv
# show logs
docker logs -f radiobrowserapi
# access api with the following link
# http://localhost:8080
# stop radiobrowser container
docker rm -f radiobrowserapi
# stop database container
docker rm -f dbserver
# remove the virtual network
docker network rm rbnet
```

### SSL

Radiobrowser does not yet support connecting with https to it directly. You have to add a reverse proxy like Apache or Nginx.

#### Apache config

```bash
# install packages (ubuntu 18.04)
sudo apt install apache2

# enable apache modules
sudo a2enmod proxy_http
sudo systemctl restart apache2
sudo mkdir -p  /var/www/radio
```

Apache config file example

```apache
<VirtualHost *:80>
    ServerName my.servername.com

    ServerAdmin webmaster@programmierecke.net
    DocumentRoot /var/www/radio

    ErrorLog ${APACHE_LOG_DIR}/error.radio.log
    CustomLog ${APACHE_LOG_DIR}/access.radio.log combined

    ProxyPass "/"  "http://localhost:8080/"
    ProxyPassReverse "/"  "http://localhost:8080/"

    <Directory /var/www/radio/>
        AllowOverride All
        Order allow,deny
        allow from all
    </Directory>
</VirtualHost>
```

Follow this guide to get a free certificate
<https://certbot.eff.org/>

### Ansible role

```bash
# clone this project
git clone https://github.com/segler-alex/radiobrowser-api-rust.git
cd radiobrowser-api-rust
# checkout stable
git checkout stable
# deploy, change email adress, for ssl with certbot
ansible-playbook -e "email=test@example.com" -e "version=0.7.24" -e "ansible_python_interpreter=auto" -i "test.example.com,test2.example.com" ansible/playbook.yml
```

## Building

### Distribution tar.gz

```bash
./builddist.sh
```

### Debian/Ubuntu package

```bash
cargo install cargo-deb
cargo deb # run this in your Cargo project directory
```

### With docker
Generate deb and tar.gz distribution with the help of docker. This has the following upsides:
* platform independent builds
* clean builds

```bash
docker run -w /root -v $(pwd):/root ubuntu:bionic bash build_with_docker.sh
```

### Kubernetes helm chart
There is a helm chart at https://gitlab.com/radiobrowser/helm-charts

## Development

### Run a test environment in multiple shells

```bash
# 1.Shell: start db
docker run -e MYSQL_DATABASE=radio -e MYSQL_USER=radiouser -e MYSQL_PASSWORD=password -e MYSQL_RANDOM_ROOT_PASSWORD=true -p 3306:3306 --rm --name dbserver mariadb --character-set-server=utf8mb4 --collation-server=utf8mb4_unicode_ci
# 2.Shell: start radiobrowser with local config
cargo run -- -f radiobrowser-dev.toml

# 3.Shell IF NEEDED: check content of database directly
docker exec -it dbserver bash
mysql -D radio -u radiouser -ppassword
```

## Environment vars / parameters
A print of the help page of the main executable. You can see parameters here. Many of them may also be set by environment variable (env: ..).
```
$ radiobrowser-api-rust --help
radiobrowser-api-rust 0.7.24
segler_alex@web.de
HTTP Rest API for radiobrowser

USAGE:
    radiobrowser-api-rust [FLAGS] [OPTIONS]

FLAGS:
        --help       Prints help information
    -v, --verbose    increases the log level. can be specified mutliple times 0..3
    -V, --version    Prints version information

OPTIONS:
    -a, --allow-database-downgrade <ALLOW_DATABASE_DOWNGRADE>
            allows downgrade of database if tables were created with newer software version

        --broken_stations_never_working_timeout <BROKEN_STATIONS_NEVER_WORKING_TIMEOUT>
            Broken streams are removed after this timespan, if they have never worked. [env:
            BROKEN_STATIONS_NEVER_WORKING_TIMEOUT=]
        --broken_stations_timeout <BROKEN_STATIONS_TIMEOUT>
            Broken streams are removed after this timespan. [env: BROKEN_STATIONS_TIMEOUT=]

        --cache-ttl <DURATION>
            time to life for cache items [env: CACHETTL=]

        --cache-type <CACHETYPE>
            one of none,builtin,redis,memcached [env: CACHETYPE=]

        --cache-url <URL>
            url to access cache server [env: CACHEURL=]

        --checks_timeout <CHECKS_TIMEOUT>
            Checks are removed after this timespan. [env: CHECKS_TIMEOUT=]

        --chunk-size-changes <CHUNK_SIZE_CHANGES>
            chunk size for downloading changes [env: CHUNK_SIZE_CHANGES=]

        --chunk-size-checks <CHUNK_SIZE_CHECKS>
            chunk size for downloading checks [env: CHUNK_SIZE_CHECKS=]

        --click_valid_timeout <CLICK_VALID_TIMEOUT>
            Possible clicks from the same IP. IPs are removed after this timespan. [env: CLICK_VALID_TIMEOUT=]

        --clicks_timeout <CLICKS_TIMEOUT>
            Clicks are removed after this timespan. [env: CLICKS_TIMEOUT=]

    -c, --concurrency <CONCURRENCY>
            streams checked in parallel [env: CONCURRENCY=]

    -f, --config-file <CONFIG-FILE>
            Path to config file [env: CONFIG_FILE=]  [default: /etc/radiobrowser.toml]

    -d, --database <DATABASE_URL>
            Database connection url [env: DATABASE_URL=]

    -x, --delete <DELETE>
            delete broken stations according to rules [env: DELETE=]

        --enable-check <ENABLE_CHECK>
            enable station checks [env: ENABLE_CHECK=]

        --favicon <FAVICON>
            check favicons and try to repair them [env: FAVICON=]

    -i, --ignore-migration-errors <IGNORE_MIGRATION_ERRORS>                                ignore errors in migrations
    -h, --host <HOST>
            listening host ip [env: HOST=]

    -p, --port <PORT>                                                                      listening port [env: PORT=]
    -l, --log-dir <LOG-DIR>
            Path to log dir [env: LOG_DIR=]

    -j, --log-json <LOG_JSON>                                                              Log in JSON format
        --max_depth <MAX_DEPTH>
            max recursive link check depth [env: MAX_DEPTH=]

    -m, --mirror <MIRROR>...
            address of other radiobrowser server to pull updates from

    -q, --mirror-pull-interval <MIRROR_PULL_INTERVAL>
            pull from mirrors at an interval [env: MIRROR_PULL_INTERVAL=]

        --pause <PAUSE>
            database check pauses [env: PAUSE=]

    -e, --prometheus-exporter <PROMETHEUS_EXPORTER>
            export statistics through a prometheus compatible exporter

        --prometheus-exporter-prefix <PROMETHEUS_EXPORTER_PREFIX>
            prefix for all exported values on /metrics

    -r, --retries <RETRIES>
            Max number of retries for station checks [env: RETRIES=]

        --server-country-code <server-country-code>
            2 letter country code for server location [env: SERVERCOUNTRYCODE=]

        --server-location <server-location>
            freeform location server string [env: SERVERLOCATION=]

    -s, --server-url <SERVER_URL>
            full server url that should be used in docs [env: SERVER_URL=]

        --source <SOURCE>
            Source string for database check entries [env: SOURCE=]

    -g, --static-files-dir <STATIC_FILES_DIR>
            directory that contains the static files [env: STATIC_FILES_DIR=]

    -n, --stations <STATIONS>
            batch size for station checks [env: STATIONS=]

        --tcp_timeout <TCP_TIMEOUT>
            tcp connect/read timeout [env: TCP_TIMEOUT=]

    -t, --threads <THREADS>
            concurrent threads used by socket [env: THREADS=]

    -u, --update-caches-interval <UPDATE_CACHES_INTERVAL>

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
## [0.7.24] 2022-05-01
### Fixed
* CLI: cleanup history did not correctly find duplicates
* CHECK: update all fields for stations
* CHECK: do check first and then local corrections
## [0.7.23] 2022-04-18
### Added
* CLI: basic cli commands for database maintenance
* API: countrycode field to countries endpoint (#148)
### Changed
* API: Deprecated countrycodes endpoint
* API: support uppercase stationuuids (#147)
### Fixed
* CHECK: Station compare and update does not create new history entries if not needed anymore

## [0.7.22] 2021-11-10
### Fixed
* CHECK: check for http result code for favicons
* PULL: insert duplicates

## [0.7.21] 2021-11-09
### Added
* CHECK: check websites for favicons, if favicon is currently empty
* CLEANUP: recheck favicon every day if still existing and delete from station if not
* CONFIG: regular automatic load/download of config
* CONFIG: reload main config on HUP and regular automatic config reload (does not get used for db connection and api, but for regular jobs)
* CONFIG: added error log level, now there are 5 log levels, 0-4
* CHECK: tags replace
* API: added identifier and image tags to xspf output

### Changed
* CLEANUP: add station change entries into the database for language and favicon updates
* DB: use transactions for migrations
* DB: allow update of existing stations with pulled changes

### Fixed
* CONFIG: replace config in shared memory

## [0.7.20] 2021-10-16
### Added
* CLEANUP: language replace, language to code
* API: iso 639 code output to language list
* DB: use binary collation for tags to fix compare
* CONFIG: allow pulling of replacement csv info from web
* CONFIG: reload csv list files on HUP signal

## [0.7.19] 2021-10-11
### Changed
* DOCKER: use alpine as base image

### Fixed
* CHECK: update stations with new languagecodes
* DB: check for empty lists in streaming servers api
* DB: collect streaming servers only if enabled

## [0.7.18] 2021-10-10
### Changed
* DB: removed fragment and query from streaming server urls
* CHECK: removed fragment and query from streaming server urls
* CLEANUP: remove unreferenced streaming server after 1 day

## [0.7.17] 2021-10-10
### Added
* API: new order type "changetimestamp"
* API: streaming servers
* DB: streaming servers
* CHECK: streaming servers

## [0.7.16] 2021-09-29
### Fixed
* DB: save geolat and geolong to checks table

## [0.7.15] 2021-09-05
### Added
* API: search by extended info marker
* API: search by is_https
* API: exposed iso_3166_2 in station lists

### Fixed
* DB: countrysubdivisioncode is now able to hold full 6 chars

### Updated
* dependencies

## [0.7.14] 2021-05-31
### Fixed
* CHECK: new av-stream-info version does fix handling of depth limit for playlists

## [0.7.13] 2021-05-02
### Added
* API: limit and offset parameters to countries, countrycodes, languages, tags, states, codecs endpoints
* PULL: automatically remove duplicates from database

## [0.7.12] 2021-04-22
### Added
* API: multiple station endpoints do not support limit,offset and hidebroken parameters
* DOCS: added missing hidebroken to advanced search

### Fixed
* API: accept numbers in json post

### Removed
* API: /stations/improvable

## [0.7.11] 2021-04-16
### Fixed
* CHECK: insert of station checks did not always use pregenerated uuids

## [0.7.10] 2021-04-16
### Added
* API: fields server_location and server_country_code to /json/config and /xml/config
* CHECK: added attribute ssl_error to /json/checks endpoint
* API: /json/checksteps, /xml/checksteps, /csv/checksteps
* API: languagecodes, geo_lat, geo_long
* CHECK: extract languagecodes, geo_lat, geo_long
* API: iso 8601 datetime fields

### Changed
* CHECK: do not ignore streams with broken ssl, just mark them
* Dependency upgrade: av-stream-info-rust 0.10.0

### Fixed
* CHECK: do not throw away important check

## [0.7.9] 2021-04-07
### Fixed
* PULL: fixed importing external extra check info

## [0.7.8] 2021-04-07
### Added
* DB: save more information from checks (ServerSoftware, Sampling, LanguageCodes, TimingMs, CountrySubdivisionCode)
* API: add more information for checks

## [0.7.7] 2021-04-05
### Added
* CLEAN: clean urls in db
### Changed
* API: check url and homepage on insert if correct
* CHECK: always ignore text/html urls

## [0.7.6] 2021-04-03
### Added
* API: limit parameter for format/station/changed
* API: limit parameter for format/checks 
* PULL: parameter chunk-size-changes
* PULL: parameter chunk-size-checks

### Fixed
* DB: database name is now allowed to be different from "radio"
* CHECK: bitrate detection of stream works now on multiple bitrate headers
* PULL: try to import stations that are not existing locally for pulled checks
* PULL: votes syncing (#111)

### Changed
* Dependency upgrade: prometheus 0.12.0
* Dependency upgrade: redis 0.20.0
* Dependency upgrade: reqwest 0.11.2
* Dependency upgrade: av-stream-info-rust 0.8.2

## [0.7.5] 2021-01-05
### Changed
* Use do-not-index header of streams to delete them from the database
* Override "Url" field in database if stream override is set
* Support additional stream headers from https://www.stream-meta.info/version_2_headers.html

### Fixed
* DOCKER: added root certificates to container

## [0.7.4] 2020-12-11
### Changed
- PKG: removed mysql dependency from deb package, this enables use of mariadb
### Fixed
* Docker build

## [0.7.3] 2020-12-08
### Added
- API: CSV output

### Changed
- API: ignore country parameter on station add. it is autogenerated from countrycode (ISO 3166-1 Alpha2).
- API: allow only countrycodes with the exact length of 2 on insert.
- API: ignore case on subtable selects (/format/codecs, /format/countries, /format/countrycodes)

src/config.rs
use clap::{App, Arg};
use std::fs;

#[derive(Debug)]
pub struct Config {
    pub listen_host: String,
    pub listen_port: i32,
    pub prometheus_exporter: bool,
    pub prometheus_exporter_prefix: String,
    pub connection_string: String,
    pub update_caches_interval: u64,
    pub ignore_migration_errors: bool,
    pub allow_database_downgrade: bool,
    pub threads: usize,
    pub server_url: String,
    pub static_files_dir: String,
    pub log_dir: String,
    pub servers_pull: Vec<String>,
    pub mirror_pull_interval: u64,
    pub log_level: usize,
    pub click_timeout_hours: u32,

    pub concurrency: usize,
    pub check_stations: u32,
    pub enable_check: bool,
    pub delete: bool,
    pub favicon: bool,
    pub pause_seconds: u64,
    pub tcp_timeout: u64,
    pub max_depth: u8,
    pub retries: u8,
    pub source: String,
    pub useragent: String,
}

fn get_option_string(
    matches: &clap::ArgMatches,
    config: &toml::Value,
    setting_name: &str,
    default_value: String,
) -> String {
    let value_from_clap = matches.value_of(setting_name);
    if let Some(value_from_clap) = value_from_clap {
        return value_from_clap.to_string();
    }

    let setting = config.get(setting_name);
    if let Some(setting) = setting {
        if setting.is_str() {
            let setting_decoded = setting.as_str();
            if let Some(setting_decoded) = setting_decoded {
                return String::from(setting_decoded);
            }
        }
    }

    default_value
}

fn get_option_number(
    matches: &clap::ArgMatches,
    config: &toml::Value,
    setting_name: &str,
    default_value: i64,
) -> i64 {
    let value_from_clap = matches.value_of(setting_name);
    if let Some(value_from_clap) = value_from_clap {
        return value_from_clap.to_string().parse().unwrap_or(default_value);
    }

    let setting = config.get(setting_name);
    if let Some(setting) = setting {
        let setting_decoded = setting.as_integer();
        if let Some(setting_decoded) = setting_decoded {
            return setting_decoded;
        }
    }

    default_value
}

fn get_option_bool(
    matches: &clap::ArgMatches,
    config: &toml::Value,
    setting_name: &str,
    default_value: bool,
) -> bool {
    let value_from_clap = matches.value_of(setting_name);
    if let Some(value_from_clap) = value_from_clap {
        return value_from_clap.to_string().parse().unwrap_or(default_value);
    }

    let setting = config.get(setting_name);
    if let Some(setting) = setting {
        let setting_decoded = setting.as_bool();
        if let Some(setting_decoded) = setting_decoded {
            return setting_decoded;
        }
    }

    default_value
}

fn get_hosts_from_config(config: &toml::Value) -> Vec<String> {
    let mut list = vec![];
    let setting = config.get("pullservers");
    if let Some(setting) = setting {
        let setting_decoded = setting.as_table();
        if let Some(setting_decoded) = setting_decoded {
            for i in setting_decoded {
                let host = i.1.get("host");
                if let Some(host) = host {
                    let host_str = host.as_str();
                    if let Some(host_str) = host_str {
                        list.push(host_str.to_string());
                    }
                }
            }
        }
    }
    list
}

pub fn load_config() -> Config {
    let hostname_str: String = hostname::get().map(|os_string| os_string.to_string_lossy().into_owned()).unwrap_or("".to_string());

    let matches = App::new("stream-check")
        .version(crate_version!())
        .author("segler_alex@web.de")
        .about("HTTP Rest API for radiobrowser")
        .arg(
            Arg::with_name("config-file")
                .short("f")
                .long("config-file")
                .value_name("CONFIG-FILE")
                .help("Path to config file")
                .env("CONFIG_FILE")
                .default_value("/etc/radiobrowser.toml")
                .takes_value(true),
        ).arg(
            Arg::with_name("log-dir")
                .short("l")
                .long("log-dir")
                .value_name("LOG-DIR")
                .help("Path to log dir")
                .env("LOG_DIR")
                .takes_value(true),
        ).arg(
            Arg::with_name("database")
                .short("d")
                .long("database")
                .value_name("DATABASE_URL")
                .help("Database connection url")
                .env("DATABASE_URL")
                .takes_value(true),
        ).arg(
            Arg::with_name("listen-host")
                .short("h")
                .long("host")
                .value_name("HOST")
                .help("listening host ip")
                .env("HOST")
                .takes_value(true),
        ).arg(
            Arg::with_name("server-url")
                .short("s")
                .long("server-url")
                .value_name("SERVER_URL")
                .help("full server url that should be used in docs")
                .env("SERVER_URL")
                .takes_value(true),
        ).arg(
            Arg::with_name("listen-port")
                .short("p")
                .long("port")
                .value_name("PORT")
                .help("listening port")
                .env("PORT")
                .takes_value(true),
        ).arg(
            Arg::with_name("prometheus-exporter")
                .short("e")
                .long("prometheus-exporter")
                .value_name("PROMETHEUS_EXPORTER")
                .takes_value(true)
                .help("export statistics through a prometheus compatible exporter"),
        ).arg(
            Arg::with_name("prometheus-exporter-prefix")
                .long("prometheus-exporter-prefix")
                .value_name("PROMETHEUS_EXPORTER_PREFIX")
                .takes_value(true)
                .help("prefix for all exported values on /metrics"),
        ).arg(
            Arg::with_name("threads")
                .short("t")
                .long("threads")
                .value_name("THREADS")
                .help("concurrent threads used by socket")
                .env("THREADS")
                .takes_value(true),
        ).arg(
            Arg::with_name("mirror")
                .short("m")
                .long("mirror")
                .value_name("MIRROR")
                .help("address of other radiobrowser server to pull updates from")
                .multiple(true)
                .takes_value(true),
        ).arg(
            Arg::with_name("update-caches-interval")
                .short("u")
                .long("update-caches-interval")
                .value_name("UPDATE_CACHES_INTERVAL")
                .help("update caches at an interval in seconds")
                .env("UPDATE_CACHES_INTERVAL")
                .takes_value(true),
        ).arg(
            Arg::with_name("mirror-pull-interval")
                .short("q")
                .long("mirror-pull-interval")
                .value_name("MIRROR_PULL_INTERVAL")
                .help("pull from mirrors at an interval in seconds")
                .env("MIRROR_PULL_INTERVAL")
                .takes_value(true),
        ).arg(
            Arg::with_name("ignore-migration-errors")
                .short("i")
                .long("ignore-migration-errors")
                .value_name("IGNORE_MIGRATION_ERRORS")
                .takes_value(true)
                .help("ignore errors in migrations"),
        ).arg(
            Arg::with_name("allow-database-downgrade")
                .short("a")
                .long("allow-database-downgrade")
                .value_name("IGNORE_MIGRATION_ERRORS")
                .takes_value(true)
                .help("allows downgrade of database if tables were created with newer software version"),
        ).arg(
            Arg::with_name("log-level")
                .short("v")
                .long("log-level")
                .value_name("LOG_LEVEL")
                .takes_value(false)
                .multiple(true)
                .help("increases the log level. can be specified mutliple times 0..3"),
        ).arg(
            Arg::with_name("static-files-dir")
                .short("g")
                .long("static-files-dir")
                .value_name("STATIC_FILES_DIR")
                .help("directory that contains the static files")
                .env("STATIC_FILES_DIR")
                .takes_value(true),
        ).arg(
            Arg::with_name("source")
                .long("source")
                .value_name("SOURCE")
                .help("Source string for database check entries")
                .env("SOURCE")
                .takes_value(true),
        )
        .arg(
            Arg::with_name("useragent")
                .long("useragent")
                .value_name("USERAGENT")
                .help("user agent value for http requests")
                .env("USERAGENT")
                .takes_value(true),
        )
        .arg(
            Arg::with_name("retries")
                .short("r")
                .long("retries")
                .value_name("RETRIES")
                .help("Max number of retries for station checks")
                .env("RETRIES")
                .takes_value(true),
        )
        .arg(
            Arg::with_name("click-timeout-hours")
                .long("click_timeout_hours")
                .value_name("CLICK_TIMEOUT_HOURS")
                .help("Possible clicks from the same IP. IPs are removed after this timespan.")
                .env("CLICK_TIMEOUT_HOURS")
                .takes_value(true),
        )
        .arg(
            Arg::with_name("max-depth")
                .long("max_depth")
                .value_name("MAX_DEPTH")
                .help("max recursive link check depth")
                .env("MAX_DEPTH")
                .takes_value(true),
        )
        .arg(
            Arg::with_name("tcp-timeout")
                .long("tcp_timeout")
                .value_name("TCP_TIMEOUT")
                .help("tcp connect/read timeout in seconds")
                .env("TCP_TIMEOUT")
                .takes_value(true),
        )
        .arg(
            Arg::with_name("pause-seconds")
                .long("pause_seconds")
                .value_name("PAUSE_SECONDS")
                .help("database check pauses in seconds")
                .env("PAUSE_SECONDS")
                .takes_value(true),
        )
        .arg(
            Arg::with_name("stations")
                .short("n")
                .long("stations")
                .value_name("STATIONS")
                .help("batch size for station checks")
                .env("STATIONS")
                .takes_value(true),
        )
        .arg(
            Arg::with_name("concurrency")
                .short("c")
                .long("concurrency")
                .value_name("CONCURRENCY")
                .help("streams checked in parallel")
                .env("CONCURRENCY")
                .takes_value(true),
        )
        .arg(
            Arg::with_name("delete")
                .short("x")
                .long("delete")
                .value_name("DELETE")
                .help("delete broken stations according to rules")
                .env("DELETE")
                .takes_value(true),
        )
        .arg(
            Arg::with_name("enable-check")
                .long("enable-check")
                .value_name("ENABLE_CHECK")
                .help("enable station checks")
                .env("ENABLE_CHECK")
                .takes_value(true),
        )
        .arg(
            Arg::with_name("favicon")
                .long("favicon")
                .value_name("FAVICON")
                .help("check favicons and try to repair them")
                .env("FAVICON")
                .takes_value(true),
        ).get_matches();

    let config_file_path: String = matches.value_of("config-file").unwrap().to_string();

    debug!("Load settings from file {}", config_file_path);
    let config = {
        let contents = fs::read_to_string(config_file_path);
        match contents {
            Ok(contents) => {
                let config = toml::from_str::<toml::Value>(&contents);
                match config {
                    Ok(config) => config,
                    Err(err) => {
                        panic!("Could not decode config file: {}", err);
                    }
                }
            }
            Err(err) => {
                error!("Could not load config file: {}", err);
                toml::from_str::<toml::Value>("").unwrap()
            }
        }
    };

    let connection_string = get_option_string(
        &matches,
        &config,
        "database",
        String::from("mysql://radiouser:password@localhost/radio"),
    );
    let static_files_dir: String = get_option_string(
        &matches,
        &config,
        "static-files-dir",
        String::from("./static/"),
    );
    let log_dir: String = get_option_string(&matches, &config, "log-dir", String::from("."));
    let listen_host: String =
        get_option_string(&matches, &config, "listen-host", String::from("127.0.0.1"));
    let listen_port: i32 = get_option_number(&matches, &config, "listen-port", 8080) as i32;

    let prometheus_exporter: bool = get_option_bool(&matches, &config, "prometheus-exporter", true);
    let prometheus_exporter_prefix: String = get_option_string(&matches, &config, "prometheus-exporter-prefix", String::from("radio_browser_"));

    let server_url: String = get_option_string(
        &matches,
        &config,
        "server-url",
        String::from("http://localhost"),
    );
    let threads: usize = get_option_number(&matches, &config, "threads", 1) as usize;
    let update_caches_interval: u64 =
        get_option_number(&matches, &config, "update-caches-interval", 0) as u64;
    let mirror_pull_interval: u64 =
        get_option_number(&matches, &config, "mirror-pull-interval", 30) as u64;
    let ignore_migration_errors: bool = get_option_bool(&matches, &config, "ignore-migration-errors", false);
    let allow_database_downgrade: bool = get_option_bool(&matches, &config, "allow-database-downgrade", false);
    let log_level: usize = matches.occurrences_of("log-level") as usize;

    let concurrency: usize = get_option_number(&matches, &config, "concurrency", 1) as usize;
    let check_stations: u32 = get_option_number(&matches, &config, "stations", 10) as u32;
    let enable_check: bool = get_option_bool(&matches, &config, "enable-check", false);
    let delete: bool = get_option_bool(&matches, &config, "delete", false);
    let favicon: bool = get_option_bool(&matches, &config, "favicon", false);
    let pause_seconds: u64 = get_option_number(&matches, &config, "pause-seconds", 10) as u64;
    let tcp_timeout: u64 = get_option_number(&matches, &config, "tcp-timeout", 10) as u64;
    let max_depth: u8 = get_option_number(&matches, &config, "max-depth", 5) as u8;
    let retries: u8 = get_option_number(&matches, &config, "retries", 5) as u8;
    let source: String = get_option_string(&matches, &config, "source", hostname_str);
    let useragent = get_option_string(&matches, &config, "useragent", String::from("stream-check/0.1"));
    let click_timeout_hours = get_option_number(&matches, &config, "click-timeout-hours", 24) as u32;
    let mut servers_pull = vec![];
    let mirrors = matches.values_of("mirror");
    if let Some(mirrors) = mirrors {
        for mirror in mirrors {
            info!("Will pull from '{}'", mirror);
            servers_pull.push(mirror.to_string());
        }
    }

    let mut servers = get_hosts_from_config(&config);
    servers_pull.append(&mut servers);

    Config {
        listen_host,
        listen_port,
        prometheus_exporter,
        prometheus_exporter_prefix,
        connection_string,
        update_caches_interval,
        ignore_migration_errors,
        allow_database_downgrade,
        mirror_pull_interval,
        servers_pull,
        threads,
        server_url,
        static_files_dir,
        log_dir,
        log_level,
        concurrency,
        check_stations,
        enable_check,
        delete,
        favicon,
        pause_seconds,
        tcp_timeout,
        max_depth,
        retries,
        source,
        useragent,
        click_timeout_hours,
    }
}
